pkg/directory/okta/provider.go

```diff
@@ -2,266 +2,172 @@ package okta
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-		// if we get a 404 on the member query, it means the group doesn't exist, so we should remove it from
-		// the cached lookup and the local groups list
-		var apiErr *APIError
-		if errors.As(err, &apiErr) && apiErr.HTTPStatusCode == http.StatusNotFound {
-			delete(p.groups, group.ID)
-			groups = append(groups[:i], groups[i+1:]...)
-			i--
-			continue
-		}
-
-		if err != nil {
-			return nil, nil, err
-		}
-		for _, u := range users {
-			userIDToGroups[u.ID] = append(userIDToGroups[u.ID], group.ID)
-			userLookup[u.ID] = u
-		}
+	err = p.sync(ctx, client)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	var users []directory.User
-	for _, u := range userLookup {
-		groups := userIDToGroups[u.ID]
-		sort.Strings(groups)
-		users = append(users, directory.User{
-			ID:          u.ID,
-			GroupIDs:    groups,
-			DisplayName: u.getDisplayName(),
-			Email:       u.Profile.Email,
+	var groups []directory.Group
+	userLookup := map[string]directory.User{}
+	for _, g := range p.groups {
+		groups = append(groups, directory.Group{
+			ID:   g.Id,
+			Name: g.Profile.Name,
 		})
+		for _, u := range p.groupMembers[g.Id] {
+			du := userLookup[u.Id]
+			du.DisplayName = getUserDisplayName(u)
+			du.Email = getUserEmail(u)
+			du.GroupIDs = append(du.GroupIDs, g.Id)
+			sort.Strings(du.GroupIDs)
+			du.ID = u.Id
+			userLookup[u.Id] = du
+		}
 	}
+
+	users := maps.Values(userLookup)
+
+	// sort groups and users
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].ID < groups[j].ID
+	})
 	sort.Slice(users, func(i, j int) bool {
 		return users[i].ID < users[j].ID
 	})
 	return groups, users, nil
 }
 
-func (p *Provider) getGroups(ctx context.Context) ([]directory.Group, error) {
-	groupURL, err := p.getURL("/api/v1/groups", func(q url.Values) {
-		q.Set("limit", strconv.Itoa(p.cfg.batchSize))
-		if p.lastUpdated != nil {
-			q.Set("filter", fmt.Sprintf(`lastUpdated gt "%[1]s" or lastMembershipUpdated gt "%[1]s"`, p.lastUpdated.UTC().Format(filterDateFormat)))
-		} else {
-			now := time.Now()
-			p.lastUpdated = &now
+func (p *Provider) sync(ctx context.Context, client *okta.Client) error {
+	var lastMembershipUpdated, lastUpdated time.Time
+	for _, g := range p.groups {
+		if g.LastMembershipUpdated != nil && g.LastMembershipUpdated.After(lastMembershipUpdated) {
+			lastMembershipUpdated = *g.LastMembershipUpdated
+		}
+		if g.LastUpdated != nil && g.LastUpdated.After(lastUpdated) {
+			lastUpdated = *g.LastUpdated
 		}
-	})
-	if err != nil {
-		return nil, err
 	}
 
-	for groupURL != "" {
-		var out []apiGroupObject
-		hdrs, err := p.apiGet(ctx, groupURL, &out)
+	var changedGroups []*okta.Group
+	if lastMembershipUpdated.IsZero() || lastUpdated.IsZero() {
+		// full sync
+		p.groups = make(map[string]okta.Group)
+		p.groupMembers = make(map[string][]okta.User)
+
+		var err error
+		changedGroups, _, err = client.Group.ListGroups(ctx, &query.Params{
+			Limit: int64(p.cfg.batchSize),
+		})
```

Comment:
does this need logic to handle pagination? something like 
```suggestion
		var err error
		var resp *okta.Response
		changedGroups, resp, err = client.Group.ListGroups(ctx, &query.Params{
			Limit: int64(p.cfg.batchSize),
		})
		if err != nil {
			return fmt.Errorf("error querying all groups: %w", err)
		}

		for resp.HasNextPage() {
			var groups []*okta.Group
			resp, err = resp.Next(ctx, &groups)
			if err != nil {
				return err
			}
			changedGroups = append(changedGroups, groups...)
		}
```

Reply:
You are correct. I misunderstood the SDK.